scalableminds/webknossos
# Add measurement tools

### URL of deployed dev instance (used for testing):
- https://addmeasurementtool.webknossos.xyz/

### Steps to test:
- Open an annotation of any dataset
- Select the new measurement tool in the toolbar (the new tool should be the right most)
- Try measuring a path. Setting path point can be done via left click. Setting the last point via double-/or right-click and terminating the measurement by hitting escape or another right click.
- Try copying the measured length from the tooltip via the copy icon.
- Start measuring and terminating it while still in the process using escape. This should not lead to any errors or unexpected behaviour. (if so, I have a regression bug :see_no_evil:)
- Select the area measurement tool. Draw an area via right click. The area should always be closed by another highlighting line to highlight that the measured area is a closed area connecting the current end and start point.
- Upon releasing the left mouse drag, the highlighted line should be gone and the connecting line should be in the normal color.
- Potential bug: see below: When hitting escape while measuring an area, the area measurement tool is not correctly reset.
- Please report any unexpected behavior. 

- Test the tracing tool. It should still work as expected as this PR fiddled around with the helper geometry used by the tracing tool.

### TODOs / Bugs:
- ~~[ ] Add fill area for area measurement tool~~ We decided against that as this is quite the hurdle and thus went for always having a highlighted connection line between the start and end of the drawn area to highlight that the user is measuring an area.
- [x] When finished marking the area the tooltip can be moved around by left clicking.


### Known issue / to discuss:
- [ ] When canceling the area measurement via pressing escape while the mouse is still pressed down and then continuing dragging the mouse, the tooltip reappears and the measurement continues. Should this be tackled or is this misuse of the tool and the user can be expected to redo the measurement? In this scenario, no exception is thrown but the measured area is not correctly canceled upon hitting escape. So keeping it as it is right now, is acceptable imo.

### Issues:
- fixes #3693

------
(Please delete unneeded items, merge only when none are left open)
